compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -135,68 +137,131 @@ fn with_fresh_ty_vars<'cx, 'tcx>(
+    if tcx.has_attr(impl1_def_id, sym::rustc_strict_coherence)
+        != tcx.has_attr(impl2_def_id, sym::rustc_strict_coherence)
+    {
+        bug!("Use strict coherence on both impls",);
+    }
+
+    if tcx.has_attr(impl1_def_id, sym::rustc_with_negative_coherence)
+        != tcx.has_attr(impl2_def_id, sym::rustc_with_negative_coherence)
+    {
+        bug!("Use with negative coherence on both impls",);
+    }
+
+    if tcx.has_attr(impl1_def_id, sym::rustc_strict_coherence) {
+        OverlapMode::Strict
+    } else if tcx.has_attr(impl1_def_id, sym::rustc_with_negative_coherence) {
+        OverlapMode::WithNegative
+    } else {
+        OverlapMode::Stable
+    }
+}
+
 /// Can both impl `a` and impl `b` be satisfied by a common type (including
 /// where-clauses)? If so, returns an `ImplHeader` that unifies the two impls.
 fn overlap<'cx, 'tcx>(
     selcx: &mut SelectionContext<'cx, 'tcx>,
     skip_leak_check: SkipLeakCheck,
-    a_def_id: DefId,
-    b_def_id: DefId,
+    impl1_def_id: DefId,
+    impl2_def_id: DefId,
 ) -> Option<OverlapResult<'tcx>> {
-    debug!("overlap(a_def_id={:?}, b_def_id={:?})", a_def_id, b_def_id);
+    debug!("overlap(impl1_def_id={:?}, impl2_def_id={:?})", impl1_def_id, impl2_def_id);
 
     selcx.infcx().probe_maybe_skip_leak_check(skip_leak_check.is_yes(), |snapshot| {
-        overlap_within_probe(selcx, skip_leak_check, a_def_id, b_def_id, snapshot)
+        overlap_within_probe(selcx, skip_leak_check, impl1_def_id, impl2_def_id, snapshot)
     })
 }
 
 fn overlap_within_probe<'cx, 'tcx>(
     selcx: &mut SelectionContext<'cx, 'tcx>,
     skip_leak_check: SkipLeakCheck,
-    a_def_id: DefId,
-    b_def_id: DefId,
+    impl1_def_id: DefId,
+    impl2_def_id: DefId,
     snapshot: &CombinedSnapshot<'_, 'tcx>,
 ) -> Option<OverlapResult<'tcx>> {
-    fn loose_check<'cx, 'tcx>(
-        selcx: &mut SelectionContext<'cx, 'tcx>,
-        o: &PredicateObligation<'tcx>,
-    ) -> bool {
-        !selcx.predicate_may_hold_fatal(o)
-    }
-
-    fn strict_check<'cx, 'tcx>(
-        selcx: &SelectionContext<'cx, 'tcx>,
-        o: &PredicateObligation<'tcx>,
-    ) -> bool {
-        let infcx = selcx.infcx();
-        let tcx = infcx.tcx;
-        o.flip_polarity(tcx)
-            .as_ref()
-            .map(|o| selcx.infcx().predicate_must_hold_modulo_regions(o))
-            .unwrap_or(false)
-    }
+    let infcx = selcx.infcx();
+    let tcx = infcx.tcx;
 
     // For the purposes of this check, we don't bring any placeholder
     // types into scope; instead, we replace the generic types with
     // fresh type variables, and hence we do our evaluations in an
     // empty environment.
     let param_env = ty::ParamEnv::empty();
 
-    let a_impl_header = with_fresh_ty_vars(selcx, param_env, a_def_id);
-    let b_impl_header = with_fresh_ty_vars(selcx, param_env, b_def_id);
+    let impl1_header = with_fresh_ty_vars(selcx, param_env, impl1_def_id);
```

Comment:
I think we can break this down a bit. We had divided things into:

* stable
* strict

But *actually* strict mode is kind of incomplete (and not very well named). All of this boils down to "ways to prove that impls are disjoint". Really there are four ways:

* "unification failure" -- cannot unify the headers for any value of the impl parameters;
* "negative impl" -- there is a negative impl that proves they are disjoint;
* "implicit negative" -- the old school way, where there is no positive impl and the type is local (or fundamental).

`stable_disjoint` is really "unification failure" OR "implicit negative", and `strict_disjoint` is 'negative impl".

The error messages seem to be expecting and relying "unification failure".

I think I would break this down. Separate "stable disjoint" into its two constituent parts, and then *always* do the "unification failure" (since that makes sense even in 'strict mode') and rewrite the logic like so:

```rust
let overlap_mode = overlap_mode(...);

if overlap_mode.use_negative_impl() {
    if negative_impl(impl1, impl2) || negative_impl(impl2, impl1) {
       return None;
    }
}

let impl1_header = with_fresh_ty_vars();
let impl2_header = with_fresh_ty_vars();
if let Err(_) = equate_impl_headers(impl1_header, impl2_header) {
    return None; // "unification failure"
}

if overlap_mode.use_implicit_negative {
    if implicit_negative(impl1_header, impl2_header, impl1, impl2) {
        return None;
    }
}

leakcheck;

Some(...);
```

Reply:
This was one of the things that I was thinking about, restructured the code in https://github.com/rust-lang/rust/pull/93175/commits/e5f2fdb53939dc01ea6cad6e369c36aaa96f6e57
Thanks Niko.